### PR TITLE
NFC Unit tests part 1.1

### PR DESCRIPTION
**Pull request by Astrrra**
# What's new

- NFC-A 7byte UID file save/load test, MFC 1/4k 4/7byte UID save/load test with default values.
- nfc_device_save[_shadow] now takes just the device and the path as an argument. **Requires thorough testing.**

# Verification 

- Run `unit_tests nfc` in the CLI. All tests up to MFC 4k7b should pass on the current firmware.
- Check all the places where nfc_device_save is used, including saving files that are already existing. 

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/debug/unit_tests/nfc/nfc_test.c [hunk 1]
```diff
@@ -17,6 +19,13 @@
 #define NFC_TEST_SIGNAL_SHORT_FILE "nfc_nfca_signal_short.nfc"
 #define NFC_TEST_SIGNAL_LONG_FILE "nfc_nfca_signal_long.nfc"
 #define NFC_TEST_DICT_PATH EXT_PATH("unit_tests/mf_classic_dict.nfc")
+#define NFC_TEST_NFCA_PATH EXT_PATH("nfc/nfca.nfc")
+#define NFC_TEST_MFC_1K_4B_PATH EXT_PATH("nfc/mf_classic_1k_4b.nfc")
+#define NFC_TEST_MFC_4K_4B_PATH EXT_PATH("nfc/mf_classic_4k_4b.nfc")
+#define NFC_TEST_MFC_1K_7B_PATH EXT_PATH("nfc/mf_classic_1k_7b.nfc")
+#define NFC_TEST_MFC_4K_7B_PATH EXT_PATH("nfc/mf_classic_4k_7b.nfc")
+#define NFC_TEST_MFU_PATH EXT_PATH("nfc/mf_ultralight.nfc")
+#define NFC_TEST_MFDES_PATH EXT_PATH("nfc/mf_desfire.nfc")
```
Comment (DrZlo13) [inline applications/debug/unit_tests/nfc/nfc_test.c]: Use the unit_tests folder for resources needed for unit testing.
Also add the unit test assets to the assets folder if they need to be provided for the unit test.

File: applications/debug/unit_tests/nfc/nfc_test.c [hunk 2]
```diff
@@ -287,9 +296,630 @@ MU_TEST(mf_classic_dict_load_test) {
...
+    memcpy(nfc->dev_data.nfc_data.uid, uid, 7);
+    nfc->dev_data.nfc_data.uid_len = 7;
+
+    nfc->dev_data.nfc_data.sak = 0x08;
+    nfc->dev_data.nfc_data.atqa[0] = 0x00;
+    nfc->dev_data.nfc_data.atqa[1] = 0x04;
+    nfc->dev_data.nfc_data.type = FuriHalNfcTypeA;
+
+    // Save the NFC device data to the file
+    mu_assert(nfc_device_save(nfc, "nfca"), "nfc_device_save == true assert failed\r\n");
```
Comment (DrZlo13) [inline applications/debug/unit_tests/nfc/nfc_test.c]: It looks like nfc_device_save is not behaving correctly, it should take the full path to the file as arguments.

**Review by DrZlo13 — DISMISSED** [review]

File: applications/debug/unit_tests/nfc/nfc_test.c (hunk 2 above)
Comment (Astrrra) [reply applications/debug/unit_tests/nfc/nfc_test.c]: It was very painful, but now it does

File: lib/nfc/nfc_device.c [hunk 3]
```diff
@@ -1073,12 +1059,16 @@ static bool nfc_device_save_file(
...
 bool nfc_device_save(NfcDevice* dev, const char* dev_name) {
-    return nfc_device_save_file(dev, dev_name, NFC_APP_FOLDER, NFC_APP_EXTENSION, true);
+    return nfc_device_save_file(dev, dev_name);
 }
 
 bool nfc_device_save_shadow(NfcDevice* dev, const char* dev_name) {
     dev->shadow_file_exist = true;
-    return nfc_device_save_file(dev, dev_name, NFC_APP_FOLDER, NFC_APP_SHADOW_EXTENSION, true);
+    // Replace extension from .nfc to .shd if necessary
+    FuriString* path = furi_string_alloc();
```
Comment (gornekich) [inline lib/nfc/nfc_device.c]: use nfc_device_get_shadow_path()

File: applications/main/nfc/scenes/nfc_scene_save_name.c [hunk 4]
```diff
@@ -62,7 +62,11 @@ bool nfc_scene_save_name_on_event(void* context, SceneManagerEvent event) {
                 nfc->dev->dev_data.nfc_data = nfc->dev_edit_data;
             }
             strlcpy(nfc->dev->dev_name, nfc->text_store, strlen(nfc->text_store) + 1);
-            if(nfc_device_save(nfc->dev, nfc->text_store)) {
+            // Create path by adding nfc/ to the dev_name or using the load path
+            FuriString* path = furi_string_alloc();
```
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_save_name.c]: Separate function for these manipulations

File: applications/main/nfc/scenes/nfc_scene_mf_classic_emulate.c [hunk 5]
```diff
@@ -48,7 +48,10 @@ bool nfc_scene_mf_classic_emulate_on_event(void* context, SceneManagerEvent even
            NFC_MF_CLASSIC_DATA_CHANGED) {
             scene_manager_set_scene_state(
                 nfc->scene_manager, NfcSceneMfClassicEmulate, NFC_MF_CLASSIC_DATA_NOT_CHANGED);
-            nfc_device_save_shadow(nfc->dev, nfc->dev->dev_name);
+            // Save shadow file
+            if(nfc->dev->load_path) {
```
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_mf_classic_emulate.c]: Always true
use furi_string_size instead

**Review by gornekich — CHANGES_REQUESTED** [review]

File: applications/main/nfc/scenes/nfc_scene_set_uid.c [hunk 6]
```diff
@@ -31,13 +31,17 @@ bool nfc_scene_set_uid_on_event(void* context, SceneManagerEvent event) {
...
                 nfc->dev->dev_data.nfc_data = nfc->dev_edit_data;
-                if(nfc_device_save(nfc->dev, nfc->dev->dev_name)) {
+                if(strcmp(nfc->dev->dev_name, "")) {
+                    // Create path by adding nfc/ to the dev_name or using the load path
+                    FuriString* path = furi_string_alloc();
+                    furi_string_printf(
+                        path, "%s/%s%s", NFC_APP_FOLDER, nfc->dev->dev_name, NFC_APP_EXTENSION);
                     scene_manager_next_scene(nfc->scene_manager, NfcSceneSaveSuccess);
                     consumed = true;
+                } else {
```
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_set_uid.c]: Can't get to save name from here. Revert changes

**Review by gornekich — APPROVED** [review]

**Review by skotopes — APPROVED** [review]

**Pull request merged**